### ISSUE TITLE
Client-update glitch on Island creation and restart

### ISSUE THREAD

**Issue by LihP** [rlf/uSkyBlock#336]
When you create an island this happens less often but when I restart it happens consistently.
The video will help explain the problem.
https://drive.google.com/open?id=0BxilLZQHdwZbOWF2OGJCcXgzMHc&authuser=0
A cool-down would be nice addition to the restart command so players cannot spam it.

**Comment by dutchy1001** [comment]
cant access the vid

**Comment by LihP** [comment]
Sorry, here https://drive.google.com/file/d/0BxilLZQHdwZbOWF2OGJCcXgzMHc/view?usp=sharing

**Comment by dutchy1001** [comment]
I have seen your problem before but not in an extreme like this. What other plugins do you have that could interfere?

I also notest that you use /home to get back to your island, this is not a default command used in USB. If you have alterd the Jar in any way that could be the cause too.

**Comment by LihP** [comment]
[18:48:09 INFO]: Plugins (6): Vault, WorldEdit, PermissionsEx, WorldGuard, ChestCommands, uSkyBlock

I used a bukkit command alias the jar has not been altered.

**Comment by dutchy1001** [comment]
ok just asking as altering can cause so many issue's and needed to eliminate that ofcourse.

**Comment by LihP** [comment]
In the video I had about 15 plugins active, but I tested with just the ones listed above and the results where the same apart from you can move around when glitched and see the block outlines because noCheat was removed. Hope it helps, I can start using the plugin once this bug is fixed and I can't wait to open the server. Thanks for all your help.

**Comment by dutchy1001** [comment]
Not sure if this is a bug on our side, as most servers dont have this issue. 
How did you create the skyworld? Is the skyworld set as default in the server.properties?(as that is a big NO NO)
it's late here so continue tomorrow (well for me atleast lol).

**Comment by rlf** [comment]
a) Home is most likely essentials (not a custom build of uSkyblock).
b) There is a cooldown on `/is restart`
c) There's a difference between 1.7.9 and 1.8 reg. client-refresh - might be what you are experiencing.

**Comment by LihP** [comment]
@dutchy1001 The skyblock world was created by itself and is called skyworld. The players spawn in a regular world called world. 
@rlf Thanks for the info, yeah I used the built in bukkit alias for the home command http://wiki.bukkit.org/Commands.yml
I guess the reset cool-down has an overrider perm that I have when testing as OP so I didn't notice it. Also I just checked the top of the config and I see the variable for it, sorry.

**Comment by rlf** [comment]
I've finally had a chance to look at the video, and it seems like the bug we experienced between 1.7.9 and 1.8 - they differ in "client-refresh" stuff, which is why we port players to spawn, and back, when they restart their island.

**Comment by LihP** [comment]
Heres is my version details, I wouldn't mind so much if it only happened when restarting an island, but it occasionally happens when new player create their island. Would a short delay between island creation and the teleport?
[13:18:39 INFO]: Name: uSkyBlock
[13:18:39 INFO]: Version: 2.3-HF1a
[13:18:39 INFO]: Description: Ultimate SkyBlock v2.3-HF1a-2e2dfe-dev
[13:18:39 INFO]: ------------------------------
[13:18:39 INFO]: Server: CraftBukkit git-Spigot-c5492f8-cf5a502 (MC: 1.8)
[13:18:39 INFO]: ------------------------------
[13:18:39 INFO]: Name: Vault
[13:18:39 INFO]: Version: 1.5.3-b37
[13:18:39 INFO]: ------------------------------
[13:18:39 INFO]: Name: WorldEdit
[13:18:39 INFO]: Version: 6.0;3342-78f975b9
[13:18:39 INFO]: ------------------------------
[13:18:39 INFO]: Name: WorldGuard
[13:18:39 INFO]: Version: 6.0.0-beta-05.1569-

**Comment by LihP** [comment]
To save opening an issue I just want to check if these is intentional. When a player leaves a group their inventory is cleared but their armour remains. I have tested being kicked from a group and this clears both armour and inventory.

**Comment by dutchy1001** [comment]
that can be set in the config, when islands are reset, not sure if it counts for when a player gets kicked

**Comment by LihP** [comment]
Sorry I forgot to mention that I have them all set to true and it does clear the armor when resetting the island only fails when a player leave a group.
  # Section about restarting your island (or accepting an invite).
  restart:
    # Clears the player's inventory on island create/restart
    clearInventory: true
    # Clears the player's armor on island create/restart
    clearArmor: true
    # Clears the player's enderchest on island create/restart
    clearEnderChest: true

**Comment by dutchy1001** [comment]
I'll have to ask @rlf how this is connected in the code.

**Comment by rlf** [comment]
To my knowledge, the clearing isn't done until you join something or create an island.

Might be wrong though.

**Comment by dutchy1001** [comment]
but still clearing the armor on kick but not on leave is the issue here, as the config states to clear all.

**Comment by LihP** [comment]
It clears everything when kicked but not when /is leave, just checked and the same is true for the enderchest. The code here only show the inventory being cleared https://github.com/rlf/uSkyBlock/blob/master/src/main/java/us/talabrek/ultimateskyblock/command/island/LeaveCommand.java

**Comment by LihP** [comment]
Can I also suggest adding a lava and fire block check to the isSafeLocation method in https://github.com/rlf/uSkyBlock/blob/master/src/main/java/us/talabrek/ultimateskyblock/uSkyBlock.java

**Comment by rlf** [comment]
Yes, it might, but players shouldn't be able to do anything once they have left the island.
If they join a new island, or create one them selves, their inventory will be cleared.

Is this an issue? Then I will try to fix it.

The suggestion about lava/fire should be posted as a standalone issue.
I think it already checks for lava (since that is not a solid block), but not sure.

**Comment by LihP** [comment]
Thanks for the reply, I can see how it would resolve itself if you only run an spawn and skyworld setup. We do have more worlds so this could become a problem and would be nice if did clear it, if not I will try to use the API to listen for that event.

We have opened the skyblock server to the public if you want to test the above issue, island reset void glitch can be replicated consitently if you want to test. (pvp.lihp.us)

I will open a ticket about the safe teleport, one other thing that I can also open a new one for it the ability to warn the player before using /is restart. Like a confirmation /is restart confirm, this will prevent  people making mistakes or being trolled by others.

**Comment by rlf** [comment]
Post em both :)

Reg. the multi-world setup, most use this, but I recommend using separate inventories (or even economies) for each world. Multiverse-Inventories can do this for you.

Esp. since players can otherwise go to another world, and mine resources to complete the challenges in skyworld.

Reg. the glitch, does it only happen on `/is restart` and/or `/is create`?
Do you use the "sendToSpawn" config parameter?
If so, where is your spawn located? (another world? another server?)

It might be caused by a no-cheat plugin of some sorts - but I doubt it.

**Comment by LihP** [comment]
Thanks for working on this issue, I am pleased to say that its resolved. I set the delay to 60 and still had the problem then I doubled it to 120 and tested a few times and didn't get glitched. I will test more tomorrow.

**Comment by rlf** [comment]
Good to hear.
Fingers crossed.

**Comment by LihP** [comment]
I might have hit the nail on the head with 120, or 6 seconds. I changed the islandTeleportDelay to 3 to match regular tp request and I started getting the glitch again, so I extended teleportDelay another 2 seconds (160) and it resolved the issue. Seems like it needs around 10 seconds overall. I Will let you know if I get any problems. Should I open a ticket for my suggested feature request for the confirmation command of restart and leave?

**Comment by dutchy1001** [comment]
thx for keeping us informed about your findings.

For your request for a confirmation command you can open a new ticket (great suggestion btw)

**Comment by LihP** [comment]
Alright will do, and I just got glitched when I used the reset while being at the island, so I extended another 2 seconds and its ok. I will continue to test.
Here are my current settings that work for my small island size.
    islandTeleportDelay: 3
    teleportDelay: 180

@dutchy1001 Thank, I will open a ticket asap.

**Comment by LihP** [comment]
Might be that time might not a factor, I appear to be able to resolve the glitch by moving, in previous the screen would jump around when moving. Anyway, its better thanks.

**Comment by rlf** [comment]
Closing this for now - since the original issue seems to have been resolved.

**Comment by LihP** [comment]
I still have the issue but it appears to be less common, sometimes the blocks show but I fall into the void.
I was testing with @dutchy1001  and they showed me their test server running 1.8.3 with very few plugins. I only tested it a few times but it never failed, should have tested it more cos mine can work consistently for awhile sometimes.

I ran some test with as few plugins as possible on my current build or 1.8 and when I get back from my 7 day trip I will test the same setup with 1.8.3.
[11:06:03 INFO]: Plugins (10): Vault, WorldEdit, PermissionsEx, WorldGuard, ChestCommands, RemoteToolkitPlugin, Essentials, EssentialsChat, EssentialsSpawn, uSkyBlock
>version
[11:06:14 INFO]: This server is running CraftBukkit version git-Spigot-c5492f8-cf5a502 (MC: 1.8) (Implementing API version 1.8-R0.1-SNAPSHOT)
>usb version
[11:06:22 INFO]: Name: uSkyBlock
[11:06:22 INFO]: Version: 2.3-HF3
[11:06:22 INFO]: Description: Ultimate SkyBlock v2.3-HF3-9fc81a-dev
[11:06:22 INFO]: ------------------------------
[11:06:22 INFO]: Server: CraftBukkit git-Spigot-c5492f8-cf5a502 (MC: 1.8)
[11:06:22 INFO]: ------------------------------
[11:06:22 INFO]: Name: Vault
[11:06:22 INFO]: Version: 1.5.3-b37
[11:06:22 INFO]: ------------------------------
[11:06:22 INFO]: Name: WorldEdit
[11:06:22 INFO]: Version: 6.0;3342-78f975b9
[11:06:22 INFO]: ------------------------------
[11:06:22 INFO]: Name: WorldGuard
[11:06:22 INFO]: Version: 6.0.0-beta-05.1569-
[11:06:22 INFO]: ------------------------------

Thanks again for your continued support and work on this project.

**Comment by rlf** [comment]
Let's keep it open then.

**Comment by LihP** [comment]
Seem like the problem might have been caused by the version of spigot 1.8.0 we where running. The problem was present in CraftBukkit git-Spigot-c5492f8-cf5a502 (MC: 1.8) and since updating to CraftBukkit git-Spigot-550ebac-7019900 (MC: 1.8) which is about 50 builds newer I have not had the issue. I did however update to 2.3-HF3b so that might have helped too. I will continue to keep an eye on it and let you know if it appears. Sorry for creating such a big issue thread for something that might have just been my fault.
